type-check.ts

```diff
@@ -375,6 +417,47 @@ export function tcExpr(env : GlobalTypeEnv, locals : LocalTypeEnv, expr : Expr<S
       } else {
         throw new TypeCheckError("method calls require an object");
       }
+    case "ternary":
+      const tExprIfTrue = tcExpr(env, locals, expr.exprIfTrue);
+      const tIfCond = tcExpr(env, locals, expr.ifcond);
+      const tExprIfFalse = tcExpr(env, locals, expr.exprIfFalse);
+      if (!equalType(tIfCond.a[0], BOOL)) {
+        throw new TypeCheckError("if condition must be a bool");
+      }
+      const exprIfTrueTyp = tExprIfTrue.a[0];
+      const exprIfFalseTyp = tExprIfFalse.a[0];
+      if (equalType(exprIfTrueTyp, exprIfFalseTyp)) {
+        return { ...expr, a: [exprIfTrueTyp, expr.a], exprIfTrue: tExprIfTrue, ifcond: tIfCond, exprIfFalse: tExprIfFalse };
+      }
+      const eitherTyp : Type = { tag: "either", left: exprIfTrueTyp, right: exprIfFalseTyp };
+      return { ...expr, a: [eitherTyp, expr.a], exprIfTrue: tExprIfTrue, ifcond: tIfCond, exprIfFalse: tExprIfFalse };
+    case "comprehension":
+      const tIterable = tcExpr(env, locals, expr.iterable);
+      const [isIterable, itemTyp] = isIterableType(tIterable.a[0])
+      if (!isIterable) {
+        throw new TypeCheckError(`Type ${tIterable.a[0]} is not iterable`);
+      }
+      if (locals.vars.has(expr.item)) {
+        throw new TypeCheckError(`id ${expr.item} already exists locally`);
```

Comment:
Is it necessary to reject an id if it's used elsewhere? Why can't you shadow the variable name?

Reply:
We'll make this change. Thanks.

Reply:
Resolved. Now iterator name is always shadowed. E.g.
```python
(a + 1 for a in ...)
(a * 2 for a in ...)
```
will generate temporary comprehension variables `compvar$a1` and `compvar$a2` in the WASM code.